src/data-model-providers/codegen/CodegenDataModelProvider_Read.cpp

```diff
@@ -102,17 +102,22 @@ DataModel::ActionReturnStatus CodegenDataModelProvider::ReadAttribute(const Data
 
     // Codegen logic specific: we accept AAI reads BEFORE server cluster interface, so that we are backwards compatible
     // in case some application installed AAI before Server Cluster Interfaces were supported
-    std::optional<CHIP_ERROR> aai_result = TryReadViaAccessInterface(
-        request, AttributeAccessInterfaceRegistry::Instance().Get(request.path.mEndpointId, request.path.mClusterId), encoder);
-    VerifyOrReturnError(!aai_result.has_value(), *aai_result);
+    const EmberAfAttributeMetadata * attributeMetadata =
+        emberAfLocateAttributeMetadata(request.path.mEndpointId, request.path.mClusterId, request.path.mAttributeId);
 
-    if (auto * cluster = mRegistry.Get(request.path); cluster != nullptr)
+    // we only allow AAI on ember-registered clusters
+    if (attributeMetadata)
     {
-        return cluster->ReadAttribute(request, encoder);
-    }
+        std::optional<CHIP_ERROR> aai_result = TryReadViaAccessInterface(
+            request, AttributeAccessInterfaceRegistry::Instance().Get(request.path.mEndpointId, request.path.mClusterId), encoder);
 
-    const EmberAfAttributeMetadata * attributeMetadata =
-        emberAfLocateAttributeMetadata(request.path.mEndpointId, request.path.mClusterId, request.path.mAttributeId);
+        VerifyOrReturnError(!aai_result.has_value(), *aai_result);
+
+        if (auto * cluster = mRegistry.Get(request.path); cluster != nullptr)
```

Comment:
But now we are skipping mRegistry on things that are not ember-registered.  My understanding is that we were aiming for a "hybrid" behavior here where you could use mRegistry for things that are not ember-registered (and certainly that's what the previous logic had)...

As in, should this look like this or some equivalent?

```
std::optional<CHIP_ERROR> aai_result;
if (attributeMetadata) {
    aai_result = TryReadViaAccessInterface(...);
}
VerifyOrReturnError(!aai_result.has_value(), *aai_result);

if (auto * cluster = mRegistry.Get(request.path); cluster != nullptr)
...
```

The edits to the "write" case seem to get this right.

Just to reiterate: this needs test coverage, for both this bug, and the bug the previous version of this had.

Reply:
I only skip if `aai_result.has_value` which means AAI retunred something.

So I skip mRegistry if AAI claims this works.
However valid point about test coverage: right now I test that "if AAI exists, it takes precedence", however I did not test that if AAI exists but ignores the read, I fall back. I will add that as an explicit test.